task/control-plane-agent/src/mgs_gimlet/host_phase2.rs

```diff
@@ -0,0 +1,285 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::vlan_id_from_sp_port;
+use core::sync::atomic::{AtomicBool, Ordering};
+use gateway_messages::{Header, Message, MessageKind, SpPort, SpRequest};
+use heapless::Vec;
+use idol_runtime::{Leased, RequestError};
+use task_control_plane_agent_api::ControlPlaneAgentError;
+use task_net_api::{Address, Ipv6Address, UdpMetadata};
+use userlib::{sys_get_timer, sys_post, TaskId, UnwrapLite};
+
+const SP_TO_MGS_MULTICAST_ADDR: Address = Address::Ipv6(Ipv6Address([
+    0xff, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x01, 0xde, 0, 1,
+]));
+const MGS_UDP_PORT: u16 = 22222;
+
+// We only support a single in-flight host phase 2 request at a time (i.e., we
+// do not support any pipelining). When the host makes a request of us (via
+// `host-sp-comms`, which calls our idol interface on its behalf), we may not
+// know _which_ MGS instance is able to provide the data it wants, and given
+// we're using UDP, we need to support a small number of retries. Our constants
+// here control how long we wait before we try "the other" MGS (assuming we've
+// picked one to start with), and then how long we wait before switching back
+// and retrying both. Assuming no successful replies from either MGS instance,
+// our timeline for a request will be:
```

Comment:
Per my usual feedback, it would be helpful to have a block diagram showing
a) Which Hubris tasks are sending data in which direction, and
b) Which interfaces on the host side are sending / receiving that data

I think the request comes from the serial port and the data comes from the network, but am not 100% sure!

Reply:
I didn't go full block diagram, but added a step-by-step in 77e52be88c6ae4563c0dccc98efaea18a2cf9728 - is that good enough?